### PR TITLE
[#134] Add drag and drop support off of actors

### PR DESCRIPTION
**Pull request by pathurs**
Adds several events for dragging of actors, this will allow developers to
implemented their own style of 'drag and drop', such as 'pulling'
or 'stretching'

New Events:
 - Enter 'enter'
 - Leave 'leave'
 - Pointer Enter 'pointerenter'
 - Pointer Leave 'pointerleave'
 - Pointer Drag Start 'pointerdragstart'
 - Pointer Drag End 'pointerdragend'
 - Pointer Drag Move 'pointerdragmove'
 - Pointer Drag Enter 'pointerdragenter'
 - Pointer Drag Leave 'pointerdragleave'

New Classes:
 - PointerDragEvent extends PointerEvent
 - GlobalCoordinates (contains Vectors for world, page, screen)

New ICapturePointerConfig Property:
 - captureDragEvents (whether to emit drag events to the actor)

Additional:
 - PointerEvent now contains original pointer object.
 - Deprecated some coordinate properties in favour of GlobalCoordinates
 property or a derivative.

Resolves: #134

### CODE REVIEW

**Comment by eonarheim** [comment]
@pathurs I'll review this today

File: src/engine/Actor.ts [hunk 1]
```diff
@@ -484,14 +486,34 @@ export class ActorImpl extends Class implements IActionable, IEvented {
       }
    }
 
+   private _capturePointerEvents: string[] = [
```
Comment (kamranayub) [inline src/engine/Actor.ts]: This is good; one thing we could do is increase static checking consistency by typing this to a string enum or string union (e.g. `PointerEvents`)

File: src/engine/Actor.ts [hunk 2]
```diff
@@ -484,14 +486,34 @@ export class ActorImpl extends Class implements IActionable, IEvented {
...
+      'pointerup', 'pointerdown', 'pointermove', 'pointerenter', 'pointerleave',
+      'pointerdragstart', 'pointerdragend', 'pointerdragmove', 'pointerdragenter', 'pointerdragleave'
+   ];
+
+   private _captureMoveEvents: string[] = [
```
Comment (kamranayub) [inline src/engine/Actor.ts]: Same comment as above

File: src/engine/Actor.ts [hunk 3]
```diff
@@ -825,7 +861,7 @@ export class ActorImpl extends Class implements IActionable, IEvented {
    /**
     * Returns the actor's [[BoundingBox]] calculated for this instant in world space.
     */
-   public getBounds() {
+   public getBounds(): BoundingBox {
```
Comment (kamranayub) [inline src/engine/Actor.ts]: 👍

File: src/engine/Algebra.ts [hunk 4]
```diff
@@ -496,4 +499,36 @@ export class Projection {
       }
       return 0;
    }
+}
+
+export class GlobalCoordinates {
```
Comment (kamranayub) [inline src/engine/Algebra.ts]: NICE 🥇

File: src/engine/Input/Pointer.ts [hunk 5]
```diff
@@ -66,35 +67,86 @@ const ScrollWheelNormalizationFactor = -1 / 40;
  */
 export class PointerEvent extends GameEvent<any> {
 
+   /** Deprecated. Use [[PointerEvent]].worldPos.x instead. */
```
Comment (kamranayub) [inline src/engine/Input/Pointer.ts]: I believe we can use @eonarheim's `@Obsolete()` decorator for all these

File: src/engine/Input/Pointer.ts [hunk 6]
```diff
@@ -260,34 +315,81 @@ export class Pointers extends Class {
...
          i = 0;
          len = this._pointerMove.length;
 
          for (i; i < len; i++) {
-            if (actor.contains(this._pointerMove[i].x, this._pointerMove[i].y, !isUIActor)) {
-               actor.eventDispatcher.emit('pointermove', this._pointerMove[i]);
+            pointerEvent = this._pointerMove[i];
+            pointer = pointerEvent.pointer;
+
+            if (actor.contains(pointer.lastWorldPos.x, pointer.lastWorldPos.y, isNotUIActor)) {
```
Comment (kamranayub) [inline src/engine/Input/Pointer.ts]: In general (not necessarily related to this PR), should we be worrying about Actor z-index too @eonarheim ?

**Comment by eonarheim** [comment]
@kamranayub I don't think we need to worry about Actor z-index for this, one should be able to handle any z-index logic in the handlers

File: src/engine/Input/Pointer.ts (hunk 5 above)
Comment (eonarheim) [reply src/engine/Input/Pointer.ts]: Please add this decorator code to deprecated things, it'll show warnings in the console if anything uses it.
```typescript
import { obsolete } from '../Util/Decorators';
...
@obsolete({message: 'PointerEvent.x will be removed in the 0.17 release', alternateMethod: 'PointerEvent.worldPos.x'})
public get x(): number {
```

In the JSDOC comment please change `Deprecated` to `@deprecated will be removed in the 0.17 release, please use PointerEvent.worldPos.x instead.` our documentation generation should pick this up and deploy the change when merged.

Comment (eonarheim) [reply src/engine/Input/Pointer.ts]: I've corrected my comment above, the JSDOC should use `@obsolete` for doc generation

**Comment by eonarheim** [comment]
@pathurs Let us know if you have any questions

* Add a [CHANGELOG.md](https://github.com/excaliburjs/Excalibur/blob/master/CHANGELOG.md) entry for this feature in the "Added section"
* Add `@obsolete` code decorator and JSDOC comment instead of `Deprecated`

```typescript
import { obsolete } from '../Util/Decorators';
...
@obsolete({message: 'PointerEvent.x will be removed in the 0.17 release', alternateMethod: 'PointerEvent.worldPos.x'})
/** @obsolete  Use [[PointerEvent]].worldPos.x instead. */
public get x(): number {
```

**Comment by pathurs** [comment]
Sorry for not replying sooner, work/family life has been busy.

I'll look into making these further changes for you.

@eonarheim @kamranayub

**Comment by eonarheim** [comment]
@pathurs No worries, totally understand there is no rush

Thanks for opening such a fantastic PR

File: src/engine/Actor.ts (hunk 1 above)
Comment (pathurs) [reply src/engine/Actor.ts]: I think most of the event strings in Events.ts could benefit from becoming Enums over types. Perhaps a proper refactor should be added as an issue?

@kamranayub @eonarheim fyi

**Comment by pathurs** [comment]
I've completed the CHANGELOG.md and @obsolete changes.

Just need a response about using Enums for PointerEvents.

File: src/engine/Actor.ts (hunk 1 above)
Comment (eonarheim) [reply src/engine/Actor.ts]: @pathurs I agree, doing a proper refactor is probably out of scope for this PR. Let's open a new issue to look at consolidating Event types as enums. At the very least we could still do a string type union something like? `type PointerEvents = 'pointerup' | 'pointerdown' | 'pointermove' ...`

Comment (eonarheim) [reply src/engine/Actor.ts]: @pathurs I missed this before in my summary, could you also update these string arrays to use a type union on lines 494 & 489? Last request I promise

**Comment by eonarheim** [comment]
@pathurs Thanks again for a really well done PR!

**Pull request merged**